### ISSUE TITLE
VillagerShops stops working after reboot if CustomNPCs is installed

### ISSUE THREAD

**Issue by azokthedefiler1** [DosMike/VillagerShops#69]
Having both CustomNPCs and VillagerShops installed will cause the shops to be unable to buy or sell after a server reboot.

Error for buying: "The items is currently out of stock"
Error for selling: "You do not have enough items"

Forge:
forge-1.12.2-14.23.5.2825-universal.jar

Mods:
spongeforge-1.12.2-2825-7.1.6.jar
CustomNPCs_1.12.2-(01Oct19).jar

Plugins:
LangSwitch-1.6.1.jar
LuckPerms-Sponge-5.0.78.jar
MegaMenus-0.7.jar
Nucleus-1.14.2-S7.1-MC1.12.2-plugin.jar
TotalEconomy-1.8.1.jar
VillagerShops-2.4.jar

To duplicate the issue, setup a new server with the mods/plugins above. Create a shop with minecraft:villager and any item. Stop/Retart the server. After the server is restarted at least one time after creating the shop, you are unable to buy/sell to the shop. Relinking the shop does nothing. Also, removing CustomNPCs afterwards does make the shop work again, it must be deleted and recreated.

**Comment by DosMike** [comment]
Ok, so it looks like custom NPCs adds NBT data for all items in containers automatically. So the items in a container indeed do not match the items in the players inventory, this behaviour is correct.
The proper way to add items in this case would be to use --filter ignore_nbt.

Since this can result in the filter to be a bit too loose and because players probably don't what to be bothered with typing that every time the question now is:
Should I add compatibility, that ignores all custom NPC data for items?

**Comment by azokthedefiler1** [comment]
I completely agree that the ignore_nbt may be too loose.

Can you ignore specific bits of NBT, or does it have to be all or nothing?

If possible, I'm thinking that the best option would be to add a config file option that ignores the CNPC data/tags specifically and have it default to true (IGNORE_CNPC_NBT=true).

*) If you have only VSHOP, then it works as intended
*) If you have VSHOP+CNPC, then it works as intended
*) If you have VSHOP+CNPC, and need the custom CNPC data, then you set the config flag to false

Is something like that possible?

**Comment by DosMike** [comment]
Well, I could theoretically store a list of NBT entries that get ignored.
I should not need a dependency on any mod for that.
If server owners are skilled enough to figure out the NBT they can add entries to that list.

**Comment by azokthedefiler1** [comment]
That could work and would give maximum flexibility too. I loaded the tellme mod and noticed that it may not technically be CNPC's, but forge that adds the tag "ForgeCaps". Once the item is in the chest, it gets the tag "customnpcs:itemscripteddata". If you take it out, it loses the CNPC tag but keeps the ForgeCaps tag.

Not sure what ForgeCaps is actually used for but being able to filter that out would solve all the problems (I think).

[nbt-info.txt](https://github.com/DosMike/VillagerShops/files/4239736/nbt-info.txt)

**Comment by DosMike** [comment]
I've seen this, I've use my [toolbox](https://ore.spongepowered.org/DosMike/Mike's-ToolBox) to check it. Although most stuff in there is for other plugins to use, the NBT inspector is quite nice to browse NBT in chat.
About the ForgeCaps, I think that's a generic wrapper, that might not do anything, but I'll test that tomorrow.

**Comment by azokthedefiler1** [comment]
Sounds good.

It might be worthwhile to flat out ignore empty tags too. I can pull an item from creative with no nbt, and once it's "touched" somewhow, forge adds an empty ForgeCaps tag to it. If it's just an empty placeholder, why add it in the first place?

**Comment by DosMike** [comment]
As soon as I try to look at the NBT, the items no longer have the customNPCs NBT... I'll keep the blacklist, but it seems it can stay empty? Also ForgeCaps or empty tags seem to be ignored automatically.

**Comment by azokthedefiler1** [comment]
That's good on the empty tags! I'll check again, maybe I did something wrong before.

The CustomNPC tags exist on items while they're are in the chest. I'm pretty sure you have to reboot too to get them added on items that never had it to begin with (like pulled from creative or /give).

Get items, create and stock shop, reboot server, check NBT on items in chest.

If you buy that item from the shop, with ignore_nbt filter, then CNPC tag gets stripped/removed and you end up with the empty ForgeCaps tag in-hand. Sell it back to the shop and CNPC tag is added back while it sits in the chest. Shops that don't have ignore_nbt will refuse to buy/sell items because they no longer match with in chest items that have the CustomNPC tags that were added after the last reboot.

Items affected are items that were already in the shop chest prior to the last reboot. I'm guessing that those tags get written either during the shutdown or startup process, which makes this bug particularly irritating. Everything works perfectly, then you reboot.

**Comment by DosMike** [comment]
Well the NBT is there, yes. And if i iterate through the container to compare ItemStacks as ItemStacks, it's not equal (because NBT differs). If I go through the container and compare the ItemStacks as DataContainer (Sponges wrapper for NBT), the customNPC data vanish and it compares just fine. I want to test if i delete that on accident somewhere else and if not, whether this is the case for all item NBT (which would be bad)

**Comment by azokthedefiler1** [comment]
Let me know if you need me to test anything. I'm sure i can muddle/google my way through a "git clone" and "grandlew" to get it built.

**Comment by DosMike** [comment]
I think I got it working now. I can also PM you my test-build on discord or something if you want.

**Comment by azokthedefiler1** [comment]
That would be perfect. I just tried to build the development branch and it failed with multiple errors.

**Comment by azokthedefiler1** [comment]
It seems that everything is working perfectly with the dev jar you sent me. I tried many different combinations of stocking items, buying, selling, and rebooting the server. Every time it worked perfectly.

I checked NBT tags between tests, on items in-hand and in-chest, before and after buying/selling, some with ForgeCaps tags and some without, and it seems that the ForgeCaps tags are being ignored completely now.

I did not modify the NBTblacklist option in the config file (I left it at "null").
I did not test with any of the vshop --filter options.
